### PR TITLE
Issue #933 - add profile reference validation to FHIRValidator

### PR DESCRIPTION
**Pull request by JohnTimm**
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

### CODE REVIEW

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 1]
```diff
@@ -37,21 +40,19 @@
...
     private FHIRValidator() { }
 
     /**
      * Validate a {@link Resource} against constraints in the base specification and
      * resource-asserted profile references or specific profile references but not both.
      *
-     * <p>Resource-asserted profiles which are unknown to the server will be ignored. Conversely,
-     * unknown profiles passed explicitly as arguments will result in a FHIRValidationException.
+     * <p>Resource-asserted profile references that are unknown, generate an issue with severity WARNING. Profile references
+     * passed in as arguments to this method, generate an issue with severity ERROR.
```
Comment (lmsurpre) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:52]:
```suggestion
     * <p>Resource-asserted profile references that are unknown to the server generate an issue with severity WARNING.
     * <p>Unknown profile references passed as arguments to this method generate an issue with severity ERROR.
```

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 2]
```diff
@@ -37,21 +40,19 @@
...
      *
-     * <p>Profile references that are passed into this method are only applicable to the outermost
-     * resource (not contained resources).
```
Comment (lmsurpre) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:54]: why remove this paragraph; that seems like useful information and the question has come up before wrt contained resources

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:52]: and the same change in each of the other variants

Comment (JohnTimm) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:52]: Updated Javadoc to reflect this language

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:54]: Added this comment back. It was removed inadvertently.

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 3]
```diff
@@ -156,26 +165,71 @@ public static FHIRValidator validator() {
...
+                        .text(string("Profile: " + url + " is not supported"))
+                        .build())
+                    .expression(string(resourceNode.path()))
+                    .build());
+            } else if (!ProfileSupport.isApplicable(profile, resourceType)) {
+                issues.add(Issue.builder()
+                    .severity(IssueSeverity.ERROR)
+                    .code(IssueType.INVALID)
+                    .details(CodeableConcept.builder()
+                        .text(string("Profile: " + url + " is not applicable to resource type: " + resourceType.getSimpleName()))
```
Comment (lmsurpre) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:221]: pretty minor, but i like to put the variable in `'` just so whitespace stands out
```suggestion
                        .text(string("Profile: '" + url + "' is not applicable to resource type: " + resourceType.getSimpleName()))
```

Comment (lmsurpre) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:221]: I guess its just a style difference though

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 4]
```diff
@@ -156,26 +165,71 @@ public static FHIRValidator validator() {
...
             this.evaluationContext = evaluationContext;
             this.includeResourceAssertedProfiles = includeResourceAssertedProfiles;
-            this.profiles = Arrays.asList(profiles);
-            this.evaluationContext.getTree().getRoot().accept(this);
-            return Collections.unmodifiableList(issues);
+            this.profiles = profiles;
         }
 
-        private void reset() {
-            issues.clear();
```
Comment (lmsurpre) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:178]: why no more reset?

Comment (JohnTimm) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:178]: We create a visitor each call to FHIRValidator.validate. No need to reset anything.

File: fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java [hunk 5]
```diff
@@ -120,8 +130,11 @@ private FHIRValidator() { }
      * Validate a resource, using an {@link EvaluationContext}, against constraints in the base specification and
      * resource-asserted profile references and/or specific profile references.
      *
-     * <p>Resource-asserted profiles which are unknown to the server will be ignored. Conversely,
-     * unknown profiles passed explicitly as arguments will result in a FHIRValidationException.
+     * <p>Resource-asserted profile references that are not available in the FHIRRegistry result in issues with severity WARNING.
+     *
+     * <p>Unknown profile references passed as arguments to this method result in issues with severity ERROR.
+     *
+     * <p>Profiles that are incompatible with the resource being validated result in issues with severity ERROR.
```
Comment (lmsurpre) [inline fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:137]:
```suggestion
     * <p>Profiles that are incompatible with the resource type being validated result in issues with severity ERROR.
```

Comment (lmsurpre) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java:137]: and the same change on all the other variants

Comment (JohnTimm) [reply fhir-validation/src/main/java/com/ibm/fhir/validation/FHIRValidator.java]: Updated Javadoc

**Pull request merged**